fix/loyaltycascade.lua

```diff
@@ -0,0 +1,108 @@
+        end
+    end
+
+    -- If the unit is a former member of your civilization, as well as now an
+    -- enemy of it, we make it become a member again.
+    if unit_former_member_index and unit_enemy_member_index then
+        local unit_name = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+        local civ_name = dfhack.TranslateName(df.global.world.entities.all[df.global.plotinfo.civ_id].name)
+
+        if unit_former_member_index > unit_enemy_member_index then
+            unit_former_member_index, unit_enemy_member_index = unit_enemy_member_index, unit_former_member_index
+        end
+
+        unit_entity_links:erase(unit_enemy_member_index)
+        unit_entity_links:erase(unit_former_member_index)
+        unit_entity_links:insert('#', df.histfig_entity_link_memberst{entity_id = df.global.plotinfo.civ_id, link_strength = 100})
+
+        dfhack.gui.showAnnouncement(([[loyaltycascade: %s is now a member of %s again]]):format(unit_name, civ_name), COLOR_WHITE)
+
+        fixed = true
+    end
+
+    for index, link in pairs(unit_entity_links) do
+        local link_type = link:getType()
+
+        if link_type ==  df.histfig_entity_link_type.FORMER_MEMBER and link.entity_id == df.global.plotinfo.group_id then
+            unit_former_member_index = index
+        end
+
+        if link_type ==  df.histfig_entity_link_type.ENEMY and link.entity_id == df.global.plotinfo.group_id then
+            unit_enemy_member_index = index
+        end
+    end
+
+    if unit_former_member_index and unit_enemy_member_index then
+        local unit_name = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+        local group_name = dfhack.TranslateName(df.global.world.entities.all[df.global.plotinfo.group_id].name)
+
+        if unit_former_member_index > unit_enemy_member_index then
+            unit_former_member_index, unit_enemy_member_index = unit_enemy_member_index, unit_former_member_index
+        end
+
+        unit_entity_links:erase(unit_enemy_member_index)
+        unit_entity_links:erase(unit_former_member_index)
+        unit_entity_links:insert('#', df.histfig_entity_link_memberst{entity_id = df.global.plotinfo.group_id, link_strength = 100})
+
+        dfhack.gui.showAnnouncement(([[loyaltycascade: %s is now a member of %s again]]):format(unit_name, group_name), COLOR_WHITE)
+
+        fixed = true
+    end
+
+    if fixed and unit.enemy.enemy_status_slot ~= -1 then
+        local status_cache = unit.enemy.enemy_status_cache
+        local status_slot = unit.enemy.enemy_status_slot
+
+        unit.enemy.enemy_status_slot = -1
+        status_cache.slot_used[status_slot] = false
+
+        for _, value in pairs(status_cache.rel_map[status_slot]) do
+            value = -1
+        end
+
+        for _, value in pairs(status_cache.rel_map) do
+            value[status_slot] = -1
+        end
+
+        if cache.next_slot > status_slot then
+            cache.next_slot = status_slot
+        end
```

Comment:
what if there were status slots taken above `status_slot`? does everything need to be moved down by one?

Reply:
Honestly I'm unsure, this is a 1:1 from the ruby script and testing is difficult without an actual cascade happening. Let me know what your gut says here.

Reply:
I'd say add a comment noting the possibility so we have a place to look if things go wonky. I can do this post merge.